src/python/pants/backend/python/rules/python_test_runner.py

```diff
@@ -25,35 +25,70 @@ class PyTestResult(TestResult):
 
 # TODO: Support deps
 # TODO: Support resources
-@rule(PyTestResult, [Select(HydratedTarget)])
-def run_python_test(target):
+@rule(PyTestResult, [Select(TransitiveHydratedTarget)])
+def run_python_test(transitive_hydrated_target):
+  target_root = transitive_hydrated_target.root
 
   # TODO: Inject versions and digests here through some option, rather than hard-coding it.
   pex_snapshot = yield Get(Snapshot, UrlToFetch("https://github.com/pantsbuild/pex/releases/download/v1.5.2/pex27",
                                                 Digest('8053a79a5e9c2e6e9ace3999666c9df910d6289555853210c1bbbfa799c3ecda', 1757011)))
 
+  all_targets = [target_root] + [dep.root for dep in transitive_hydrated_target.dependencies]
+
+  # Produce a pex containing pytest and all transitive 3rdparty requirements.
+  all_requirements = []
+  for maybe_python_req_lib in all_targets:
+    # This is a python_requirement().
+    if hasattr(maybe_python_req_lib.adaptor, 'requirement'):
+      all_requirements.append(str(maybe_python_req_lib.requirement))
+    # This is a python_requirement_library(). TODO: see if this is correct?!
+    if hasattr(maybe_python_req_lib.adaptor, 'requirements'):
+      for py_req in maybe_python_req_lib.adaptor.requirements:
+        all_requirements.append(str(py_req.requirement))
+
   # TODO: This should be configurable, both with interpreter constraints, and for remote execution.
   python_binary = sys.executable
```

Comment:
This is the problem Danny, or at least one of them. Pants still runs under python2.7 always. For the python3 integration failures, you're getting requirements pexes built with py27 deps. Easy to check me by looking at the pex you found containing the deps in detail. What do the wheel names say in this regard?

Reply:
The debuggability issue I was mentioning above was that I wasn't able to find said pex, which Stu has been able to assist with by pointing me to the option I probably could have found or hacked around if I had spent > 5 seconds on it. Pants running under py2.7 always makes sense and I hadn't considered that, nor did I think about the implicit version/etc of `sys.executable` at all (will do so in the future). The issue I was seeing locally warranting the complaining about debugging was showing up in the python 2 travis integration shards as well, e.g. https://travis-ci.org/pantsbuild/pants/jobs/467312557 -- I actually hadn't tried py3 at all yet, and was mostly complaining because of the engine dep graph cycle bit which I didn't try to look into yet (but I will now).

Reply:
I'm confused. What does this mean then?

> WIP because trying to run with 3rdparty deps fails during the pytest run without being able to find the 3rdparty deps for no reason I can figure out right now (the requirements definitely do make it into the pex generation).

Reply:
It was pretty objectively unclear what I was saying -- will try to avoid making people have to figure it out next time. It's not hard to leave a tab open to github until I have the time to write everything out.

Reply:
Keep in mind `always running with Py2` will hopefully no longer be true by end of the year. We're getting much closer to finishing out these remaining integration test failures.